### ISSUE TITLE
gdn: asymmetric value/key heads (num_value_heads > num_key_heads) silently produces wrong output

### ISSUE THREAD

**Issue by ohdearquant** [ohdearquant/lattice#262]
## Summary

`gated_delta_net_step` only correctly handles **symmetric** GDN head configs (`num_value_heads == num_key_heads`). For any **asymmetric** config (`num_value_heads > num_key_heads` — the general Qwen3-Next case, and the code's own default), it silently writes only the first `num_key_heads` value heads and feeds stale buffer data through the output projection, yielding wrong logits with no error.

**Not triggered by the shipped qwen3.5-0.8b** (configured 16k/16v symmetric — that's why e2e-parity is green). This is a latent gap that would silently corrupt output for larger Qwen3-Next models and for any config that omits the value-heads field.

## Evidence (crates/inference/src/attention/gdn.rs)

- State is sized for **value** heads: `s_matrices: vec![0.0; value_heads * key_dim * value_dim]` (`GatedDeltaNetState::new`, ~L87, via `cfg.linear_num_value_heads()`).
- Output width is **value** heads: `linear_output_dim() = linear_num_value_heads() * linear_value_head_dim` (qwen35_config.rs:439).
- But the step loop iterates **key** heads: `let num_heads = cfg.linear_num_key_heads;` (gdn.rs:193), `for h in 0..num_heads { ... }` (L267 recurrence, L333 gated RMSNorm).
- Net effect for `value_heads > key_heads`: steps 7-8 write `gated_norm_buf[0 .. key_heads*value_dim]`; `out_proj` then reads `gated_norm_buf[..output_dim]` = `[.. value_heads*value_dim]` (L346-347). The tail `[key_heads*value_dim .. value_heads*value_dim]` is **stale/uninitialized scratch** → wrong output.

## The footgun default

`linear_num_value_heads()` returns `self.linear_num_value_heads.unwrap_or(32)` (qwen35_config.rs:378) while `linear_num_key_heads` defaults to 16. So a config.json that omits the value-heads field gets **32 value / 16 key** — asymmetric — and silently hits the corruption path. (The hardcoded `qwen35_2b()` / qwen3.5-0.8b defaults set `Some(16)`, dodging it.)

## Misleading test coverage

`test_s_matrix_allocation_asymmetric` and `test_output_dim_asymmetric` (gdn.rs:817, 844) set `value_heads=48` and assert **only the allocation size and output_dim arithmetic** — neither runs `gated_delta_net_step` or checks its output. `test_k_head_mapping_identity` (L807) only validates the `ratio = 16/16 = 1` identity case. So the asymmetric output path is allocation-tested but never correctness-tested; the half-implemented `k_head_mapping` concept was never wired into the step loop.

## Fix sketch (needs real-model validation)

The step loop should iterate **value** heads and map each value head `h_v` to its key head `h_k = h_v / (num_value_heads / num_key_heads)` (GQA-style head sharing: q/k come from the key head, v and the recurrent state are per-value-head). Concretely:
- `for h_v in 0..num_value_heads { let h_k = h_v / ratio; /* q,k from h_k slice; v, S, output from h_v */ }`
- Verify the exact q/k sharing convention against HF `Qwen3NextGatedDeltaNet` / FLA `fused_recurrent_gated_delta_rule`.

**Validation requirement:** qwen3.5-0.8b is symmetric and cannot exercise this path. A fix must be differential-tested (max-diff < 1e-4) against HF transformers on a real **asymmetric** Qwen3-Next checkpoint (or a constructed asymmetric reference), per the repo's differential-test-first rule. Do not land a self-consistent-but-unvalidated rewrite of an 18/24-layer recurrence.

## Secondary (low): conv kernel-size underflow

`buf_len = cfg.linear_conv_kernel_dim - 1` (gdn.rs:84, and `kernel_size - 1` at ~L368) underflows `usize` if `linear_conv_kernel_dim == 0` (malformed config) → huge allocation / OOB. Cheap defensive guard: validate `linear_conv_kernel_dim >= 1` at config load (or `checked_sub`).

## Provenance

Found via an adversarial correctness sweep of gdn.rs, TBV-confirmed against source. (The same sweep flagged the decay-before-correction recurrence order as "critical" — that one is a **false positive**: the code correctly implements the *gated* delta rule, decay-before-correction, matching Qwen3-Next/FLA/HF, empirically confirmed by green e2e-parity and the 4.77e-6 real-model GDN forward gate. The sweep assumed the un-gated DeltaNet convention.)

**Comment by ohdearquant** [comment]
### Blast-radius note (relevant to *how* this gets fixed, not just the fix)

While scoping a possible interim **fail-closed guard** (reject asymmetric configs at load with `UnsupportedModel` rather than silently corrupt), I found the asymmetric configs are not hypothetical — both built-in presets are genuinely asymmetric and are exercised widely:

- `qwen36_35b_a3b()` — `linear_num_key_heads: 16`, `linear_num_value_heads: Some(32)` (2:1)
- `qwen36_27b()` — `linear_num_key_heads: 16`, `linear_num_value_heads: Some(48)` (3:1)

Used by (grep `qwen36_27b|qwen36_35b_a3b` under `crates/`):
- **7 bench examples** construct `qwen36_27b()`: `bench_concurrent`, `bench_q4_prefill`, `bench_persistent_state`, `bench_gdn_decode`, `bench_pruning`, `bench_stability`, `bench_quality`.
- **`chat_metal` binary** falls back to `qwen36_27b()` when no `config.json` is present (`src/bin/chat_metal.rs:483`).
- quarot tests (`rmsnorm_fusion.rs:428`, `forward_equivalence.rs:1285`), loading tests (`model/qwen35/loading.rs:351/370`), config/tensor-name tests use both presets.

Two consequences:

1. **A blanket fail-closed guard is *not* a bounded change.** Placed at model/state construction it would break all 7 benches, the `chat_metal` no-config fallback, and several tests. So the interim-guard option is itself a design decision (do we want lattice to *refuse* Qwen3.6 rather than run it wrongly?), not a quick patch. If a guard is wanted, it likely belongs on the **real-checkpoint load path** (`from_config_json`) only, leaving preset-driven benches alone — but that's a deliberate call.

2. **The qwen36_27b benches currently time a silently-corrupt forward pass.** They measure latency only (no output validation), so their tok/s numbers are fine *as latency*, but the GDN output for value-heads `16..47` is stale scratch. Do not ever read those benches as quality/correctness signals for the 27B shape, and don't be surprised that they "pass" today — the corruption is invisible to a latency-only bench.

Net: the real fix (iterate value-heads with q/k ratio mapping, per the fix sketch above, in **both** `gdn.rs` and the fused Metal path `gdn_fused.rs`) still gates on an asymmetric reference checkpoint + HF differential test. This note just records that the "cheap fail-closed stopgap" turns out not to be cheap, and that the existing 27B benches are not correctness evidence.

**Comment by ohdearquant** [comment]
> _Comment by Claude (Anthropic agent) on behalf of @ohdearquant._

## TBV refinement: production routes through the FUSED path, which already implements the fix sketch — this is an oracle gap, not a production-inference corruption

A fresh adversarial sweep of the GDN forward path re-surfaced this issue. Tracing the dispatch downgrades its real-world severity.

### Production never calls the buggy scalar `gated_delta_net_step`

Every production forward path calls the **fused** entry `gated_delta_net_step_fused` (or its dtype mirrors), never the scalar `gated_delta_net_step`:

- `model/qwen35/forward.rs:141` → `gated_delta_net_step_fused` (bf16 main forward)
- `forward/cpu_f16.rs:696` → `gated_delta_net_step_fused_f16`
- `forward/cpu_q8.rs:531` → `gated_delta_net_step_fused_q8`

The scalar `gated_delta_net_step` (the one this issue identifies as iterating `0..num_key_heads`) has exactly two callers, **both tests**:
- `gdn.rs:786` — the zero-weights→zero-output unit test
- `gdn_fused.rs` test module (L1058–L1485) — the scalar-vs-fused parity oracle

### The fused path already implements this issue's fix sketch

`gated_delta_net_step_fused` (gdn_fused.rs:461–465) iterates **value** heads and maps each to its key head exactly as the fix sketch prescribes:

```rust
for h in 0..value_heads {          // value-head loop, not key-head
    let k_head = h / ratio;        // ratio = value_heads / num_key_heads
    let q_start = k_head * key_dim; // q/k from the shared key head
    let v_start = v_offset + h * value_dim; // v + state + output per value head
```

So for an asymmetric Qwen3-Next checkpoint, production inference routes through the structurally-correct fused loop. The "silently corrupt output for larger Qwen3-Next models" framing in the issue body applies to the scalar reference, not to what the engine actually runs.

### Corrected scope of remaining work

1. **Scalar oracle** (`gdn.rs` `gated_delta_net_step`): still wrong for asymmetric — it iterates key heads (L267, L333). This breaks the scalar-vs-fused parity test for any asymmetric config (the parity tests currently only exercise the symmetric 16/16 case, so it's latent). Fix = mirror the fused loop (`0..value_heads`, `k_head = h/ratio`).
2. **Validation requirement transfers to the fused path.** The fused asymmetric logic is structurally correct but has **never been differential-tested** against HF `Qwen3NextGatedDeltaNet` on a real asymmetric checkpoint (qwen3.5-0.8b is symmetric). The max-diff<1e-4 requirement in this issue should target the **fused** path; the scalar fix then just needs to match the (validated) fused output.
3. **Divisibility guard** (`gdn_fused.rs:369`): `debug_assert_eq!(value_heads % num_heads, 0, …)` is stripped in release. A malformed config (`value_heads < num_keys` → `ratio = 0`, or non-divisible) currently fails via an integer div-by-zero / OOB-index panic at L462/L477 rather than a named error. Already fail-loud (not silent), so low priority — but the eventual fix should promote this to a release-active `assert!` in both paths, matching the `apply_gqa_attention` / `decode_attention_scores` convention (#318/#319).

### Unchanged

The `linear_num_value_heads().unwrap_or(32)` vs `linear_num_key_heads` default-16 footgun (issue body) is real and worth fixing regardless — an omitted value-heads field still yields an asymmetric config that the scalar oracle handles wrong.

**Net:** still gated pending an asymmetric checkpoint + HF differential validation (per the repo's differential-test-first rule), but reframed from "live production corruption" to "oracle correctness + un-validated fused asymmetric path." No production logits are currently wrong on the shipped symmetric model.
